File: EduSphereApp/edusphere-web/backend/src/main/java/com/edusphere/controller/QuizController.java
package com.edusphere.controller;

import com.edusphere.config.JwtAuthFilter.UserPrincipal;
import com.edusphere.dto.QuizSubmitRequest;
import com.edusphere.model.Certificate;
import com.edusphere.model.QuizResult;
import com.edusphere.model.User;
import com.edusphere.model.UserProgress;
import com.edusphere.repository.*;
import jakarta.transaction.Transactional;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.*;

@RestController
@RequestMapping("/api/quiz")
public class QuizController {

    private final QuizResultRepository quizResultRepository;
    private final UserProgressRepository userProgressRepository;
    private final TopicCompletionRepository topicCompletionRepository;
    private final CertificateRepository certificateRepository;
    private final UserRepository userRepository;

    public QuizController(
            QuizResultRepository quizResultRepository,
            UserProgressRepository userProgressRepository,
            TopicCompletionRepository topicCompletionRepository,
            CertificateRepository certificateRepository,
            UserRepository userRepository
    ) {
        this.quizResultRepository = quizResultRepository;
        this.userProgressRepository = userProgressRepository;
        this.topicCompletionRepository = topicCompletionRepository;
        this.certificateRepository = certificateRepository;
        this.userRepository = userRepository;
    }

    private Long getAuthenticatedUserId() {
        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        if (principal instanceof UserPrincipal) {
            return ((UserPrincipal) principal).getId();
        }
        throw new IllegalStateException("User principal not found in security context");
    }

    private int getRequiredTopicsForSubject(String subject) {
        switch (subject) {
            case "Quantum Physics": return 12;
            case "Human Anatomy": return 15;
            case "Organic Chemistry": return 10;
            case "Ancient History": return 8;
            case "Calculus III": return 14;
            case "Astrophysics": return 11;
            case "Marine Biology": return 9;
            case "Computer Science": return 20;
            case "Artificial Intelligence": return 18;
            case "Psychology 101": return 10;
            default: return 10;
        }
    }

    @PostMapping("/submit")
    @Transactional
    public ResponseEntity<?> submitQuizResult(@Valid @RequestBody QuizSubmitRequest request) {
        Long userId = getAuthenticatedUserId();
        int percentage = Math.round(((float) request.getScore() / request.getTotal()) * 100);

        QuizResult result = new QuizResult(
                userId,
                request.getSubjectName(),
                request.getTopicName(),
                request.getScore(),
                request.getTotal(),
                percentage
        );
        quizResultRepository.save(result);

        // Update progress statistics
        UserProgress progress = userProgressRepository.findByUserId(userId)
                .orElseGet(() -> new UserProgress(userId));
        progress.setTotalQuizzes(progress.getTotalQuizzes() + 1);
        progress.setTotalQuizScore(progress.getTotalQuizScore() + percentage);
        userProgressRepository.save(progress);

        Map<String, Object> body = new HashMap<>();
        body.put("message", "Quiz result saved successfully");
        body.put("score", request.getScore());
        body.put("percentage", percentage);
        return ResponseEntity.ok(body);
    }

    @GetMapping("/performance")
    public ResponseEntity<List<Map<String, Object>>> getPerformance() {
        Long userId = getAuthenticatedUserId();
        List<QuizResult> results = quizResultRepository.findByUserId(userId);

        // Group by subject and find highest score
        Map<String, Integer> subjectScores = new HashMap<>();
        for (QuizResult qr : results) {
            int currentMax = subjectScores.getOrDefault(qr.getSubjectName(), -1);
            if (qr.getPercentage() > currentMax) {
                subjectScores.put(qr.getSubjectName(), qr.getPercentage());
            }
        }

        List<Map<String, Object>> list = new ArrayList<>();
        for (Map.Entry<String, Integer> entry : subjectScores.entrySet()) {
            Map<String, Object> map = new HashMap<>();
            map.put("subject_name", entry.getKey());
            map.put("score", entry.getValue());
            list.add(map);
        }
        return ResponseEntity.ok(list);
    }

    @GetMapping("/leaderboard")
    public ResponseEntity<List<Map<String, Object>>> getLeaderboard() {
        List<UserProgress> list = userProgressRepository.findAllByOrderByStreakDesc();
        List<Map<String, Object>> leaderboard = new ArrayList<>();

        for (int i = 0; i < Math.min(list.size(), 20); i++) {
            UserProgress up = list.get(i);
            Optional<User> userOpt = userRepository.findById(up.getUserId());
            if (userOpt.isPresent()) {
                User u = userOpt.get();
                Map<String, Object> entry = new HashMap<>();
                entry.put("rank", i + 1);
                entry.put("name", u.getName());
                entry.put("streak", up.getStreak());
                entry.put("total_completed", up.getTotalCompleted());
                leaderboard.add(entry);
            }
        }
        return ResponseEntity.ok(leaderboard);
    }

    @PostMapping("/certificate")
    @Transactional
    public ResponseEntity<?> generateCertificate(@RequestBody Map<String, String> bodyRequest) {
        Long userId = getAuthenticatedUserId();
        String subject = bodyRequest.get("subject_name");
        
        if (subject == null || subject.isEmpty()) {
            Map<String, String> error = new HashMap<>();
            error.put("message", "Subject name is required");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error);
        }

        // Check if certificate already exists
        boolean certExists = certificateRepository.existsByUserIdAndSubjectName(userId, subject);
        if (certExists) {
            Map<String, String> error = new HashMap<>();
            error.put("message", "Certificate already issued for this subject");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error);
        }

        // Verify completions matching target count
        long completedCount = topicCompletionRepository.countByUserIdAndSubjectName(userId, subject);
        int requiredCount = getRequiredTopicsForSubject(subject);

        if (completedCount < requiredCount) {
            Map<String, String> error = new HashMap<>();
            error.put("message", "Incomplete subject: Completed " + completedCount + "/" + requiredCount + " topics");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error);
        }

        // Generate unique code
        String uniqueCode = "CERT-" + subject.substring(0, 2).toUpperCase() + "-" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
        Certificate certificate = new Certificate(userId, subject, uniqueCode);
        certificateRepository.save(certificate);

        Map<String, Object> body = new HashMap<>();
        body.put("message", "Certificate issued successfully");
        body.put("certificate_code", uniqueCode);
        body.put("subject_name", subject);
        return ResponseEntity.ok(body);
    }

    @GetMapping("/certificates")
    public ResponseEntity<List<Map<String, Object>>> getCertificates() {
        Long userId = getAuthenticatedUserId();
        List<Certificate> certs = certificateRepository.findByUserId(userId);
        
        List<Map<String, Object>> result = new ArrayList<>();
        for (Certificate c : certs) {
            Map<String, Object> map = new HashMap<>();
            map.put("subject_name", c.getSubjectName());
            map.put("certificate_code", c.getCertificateCode());
            map.put("issued_at", c.getIssuedAt());
            result.add(map);
        }
        return ResponseEntity.ok(result);
    }
}

File: edusphere-web/backend/controllers/quizController.js
const db = require('../config/db');
const crypto = require('crypto');

// 1. Submit quiz results
exports.submitQuizResult = async (req, res) => {
  const { subject_name, topic_name, score, total } = req.body;
  const userId = req.user.id;

  if (!subject_name || !topic_name || score === undefined || !total) {
    return res.status(400).json({ message: 'All fields are required' });
  }

  // Server-side validation — don't trust client-supplied scores blindly
  const parsedScore = parseInt(score, 10);
  const parsedTotal = parseInt(total, 10);

  if (isNaN(parsedScore) || isNaN(parsedTotal)) {
    return res.status(400).json({ message: 'Score and total must be valid numbers' });
  }

  if (parsedTotal < 1 || parsedTotal > 50) {
    return res.status(400).json({ message: 'Total must be between 1 and 50' });
  }

  if (parsedScore < 0 || parsedScore > parsedTotal) {
    return res.status(400).json({ message: 'Score must be between 0 and total' });
  }

  const percentage = Math.round((parsedScore * 100) / parsedTotal);

  const connection = await db.getConnection();
  try {
    await connection.beginTransaction();

    // Insert quiz result
    await connection.query(
      'INSERT INTO quiz_results (user_id, subject_name, topic_name, score, total, percentage) VALUES (?, ?, ?, ?, ?, ?)',
      [userId, subject_name, topic_name, parsedScore, parsedTotal, percentage]
    );

    // Update aggregate progress stats
    await connection.query(
      'UPDATE user_progress SET total_quizzes = total_quizzes + 1, total_quiz_score = total_quiz_score + ? WHERE user_id = ?',
      [percentage, userId]
    );

    await connection.commit();
    connection.release();

    res.status(200).json({
      message: 'Quiz result submitted successfully',
      percentage,
      score: parsedScore,
      total: parsedTotal
    });
  } catch (error) {
    await connection.rollback();
    connection.release();
    console.error('Submit quiz result error:', error);
    res.status(500).json({ message: 'Internal Server Error' });
  }
};

// 2. Get quiz performance statistics
exports.getPerformance = async (req, res) => {
  const userId = req.user.id;
  try {
    // Subject wise highest quiz score
    const [rows] = await db.query(
      'SELECT subject_name, MAX(percentage) as highest_score, COUNT(*) as attempts FROM quiz_results WHERE user_id = ? GROUP BY subject_name',
      [userId]
    );

    res.status(200).json(rows);
  } catch (error) {
    console.error('Get performance stats error:', error);
    res.status(500).json({ message: 'Internal Server Error' });
  }
};

// 3. Get leaderboard ranks
exports.getLeaderboard = async (req, res) => {
  try {
    const [rows] = await db.query(
      `SELECT u.id, u.name, up.streak, up.total_completed, up.total_quizzes, 
              CASE WHEN up.total_quizzes > 0 THEN ROUND(up.total_quiz_score / up.total_quizzes) ELSE 0 END as avg_score
       FROM users u
       JOIN user_progress up ON u.id = up.user_id
       WHERE u.role != 'admin'
       ORDER BY up.total_completed DESC, avg_score DESC, up.streak DESC
       LIMIT 10`
    );
    res.status(200).json(rows);
  } catch (error) {
    console.error('Get leaderboard error:', error);
    res.status(500).json({ message: 'Internal Server Error' });
  }
};

// 4. Generate Certificate for Subject Completion
exports.generateCertificate = async (req, res) => {
  const { subject_name } = req.body;
  const userId = req.user.id;

  if (!subject_name) {
    return res.status(400).json({ message: 'Subject name is required' });
  }

  // Get total topics for subject (matching Android lists)
  const totalTopicsMap = {
    'Quantum Physics': 12,
    'Human Anatomy': 15,
    'Organic Chemistry': 10,
    'Ancient History': 8,
    'Calculus III': 14,
    'Astrophysics': 11,
    'Marine Biology': 9,
    'Computer Science': 20,
    'Artificial Intelligence': 18,
    'Psychology 101': 10
  };

  const requiredCount = totalTopicsMap[subject_name] || 10;

  try {
    // Count user's completed topics in this subject
    const [completions] = await db.query(
      'SELECT COUNT(*) as count FROM topic_completions WHERE user_id = ? AND subject_name = ?',
      [userId, subject_name]
    );

    const completedCount = completions[0].count;

    if (completedCount < requiredCount) {
      return res.status(400).json({
        message: `Subject incomplete. You completed ${completedCount}/${requiredCount} topics. Master all topics to unlock the certificate!`
      });
    }

    // Check if certificate already exists
    const [existing] = await db.query(
      'SELECT * FROM certificates WHERE user_id = ? AND subject_name = ?',
      [userId, subject_name]
    );

    if (existing.length > 0) {
      return res.status(200).json({
        message: 'Certificate already unlocked!',
        certificate: existing[0]
      });
    }

    // Generate unique certificate verification code — 128-bit entropy (16 bytes)
    const certCode = 'CERT-' + crypto.randomBytes(16).toString('hex').toUpperCase();

    const [result] = await db.query(
      'INSERT INTO certificates (user_id, subject_name, certificate_code) VALUES (?, ?, ?)',
      [userId, subject_name, certCode]
    );

    res.status(201).json({
      message: 'Congratulations! Certificate unlocked successfully.',
      certificate: {
        id: result.insertId,
        user_id: userId,
        subject_name,
        certificate_code: certCode,
        issued_at: new Date()
      }
    });
  } catch (error) {
    console.error('Generate certificate error:', error);
    res.status(500).json({ message: 'Internal Server Error' });
  }
};

// 5. Get all certificates for user
exports.getCertificates = async (req, res) => {
  const userId = req.user.id;
  try {
    const [list] = await db.query(
      'SELECT id, subject_name, certificate_code, issued_at FROM certificates WHERE user_id = ? ORDER BY issued_at DESC',
      [userId]
    );
    res.status(200).json(list);
  } catch (error) {
    console.error('Get certificates error:', error);
    res.status(500).json({ message: 'Internal Server Error' });
  }
};
